Add version catalog

oguzkocer opened the pull request:
This PR integrates [android-dependency-catalog](https://github.com/Automattic/android-dependency-catalog/) which is our [version-catalog](https://docs.gradle.org/current/userguide/platforms.html) project. Before reviewing this PR, I highly suggest taking a look at the [Gradle documentation for version-catalogs](https://docs.gradle.org/current/userguide/platforms.html) and reading the [documentation](https://github.com/Automattic/android-dependency-catalog/blob/trunk/README.md) of [android-dependency-catalog](https://github.com/Automattic/android-dependency-catalog/) project.

**TL;DR:** Version-catalog is Gradle's new feature that centralizes declaration of dependencies. It uses a `TOML` file and it can be either added per project, or centralized as a Maven project which is our preferred approach since we have a great number of projects. It makes it very easy to use consistent version numbers across projects, override the centralized version numbers, use a shared dependency cache and it has type-safe auto complete for dependencies.



![auto-complete](https://user-images.githubusercontent.com/662023/168902369-e8c96b1f-119c-4cd8-bc63-84b0f77baa52.gif)



---

The first version of [android-dependency-catalog](https://github.com/Automattic/android-dependency-catalog/) was created mostly from FluxC dependencies but also includes some WCAndroid dependencies which are currently ahead of the FluxC counterpart. That's why dependencies such as `androidx-annotation`, `dagger`, `mockito` etc are overridden in `settings.gradle` file, where the version catalog is declared. As an example, [`android-dependency-catalog`'s `1.0.0` version](https://github.com/Automattic/android-dependency-catalog/blob/1.0.0/libs.versions.toml#L44) is `2.40.5` which [matches the version from WCAndroid](https://github.com/woocommerce/woocommerce-android/blob/9.2-rc-1/settings.gradle#L3) whereas [the version from current FluxC `trunk`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/f84fa97de87eff5d1f8d5c2e01b4dc217fc1ec8b/build.gradle#L73) is `2.29.1` which is why [the catalog value is overridden to `2.29.1` here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/01fd695099dcd502a5a466c6afb301d10a8e8889/settings.gradle#L55).

Hopefully this ^ showcases how version overriding works and how we'll be able to keep a centralized catalog which will make it easier to use consistent version numbers without forcing any project to any specific version before it's ready.

---

**To test**

In this PR the goal is to replace the existing dependency declarations without changing them, so I used the approach below to ensure that the current `trunk` and this PR has the exact same dependencies.

1. Switch to current `trunk` (at the time of these instructions, it's at `d0a9a1baf082e7428b6e6e6ca1c5540d5edeca74`)
2. Run `./gradlew -q dependencies :example:dependencies :fluxc:dependencies :fluxc-annotations:dependencies :fluxc-processor:dependencies :plugins:dependencies :plugins:woocommerce:dependencies :tests:dependencies :tests:api:dependencies > ~/dependencies-trunk.log`
3. Switch to this PR
4. Run `./gradlew -q dependencies :example:dependencies :fluxc:dependencies :fluxc-annotations:dependencies :fluxc-processor:dependencies :plugins:dependencies :plugins:woocommerce:dependencies :tests:dependencies :tests:api:dependencies > ~/dependencies-pr.log`
5. Compare the two logs: `git diff ~/dependencies-trunk.log ~/dependencies-pr.log`
6. Verify that there are no dependency changes

Note that if Gradle ends up running the tasks in a different order, the logs might shift around. I have a more involved script to check each module individually, but since this approach seems to work for now, I'll refrain from sharing that. If you have any issues with the comparison, please let me know and I'll share more detailed instructions.

---

Please do not merge this PR until we get a chance to collect feedback.

oguzkocer committed:
Add shared dependency catalog as 'sharedLibs'

oguzkocer committed:
Use sharedLibs for fluxc-process & tests/api

oguzkocer committed:
Use sharedLibs for fluxc module

oguzkocer committed:
Use sharedLibs for example module

oguzkocer committed:
Use sharedLibs for plugins:woocommerce module

oguzkocer committed:
Re-add encrypted log deps and change appcompat version to 1.0.2

oguzkocer committed:
Remove specific junit version and update catalog version

oguzkocer committed:
Use sharedLibs for plugins

oguzkocer committed:
Merge remote-tracking branch 'origin/trunk' into add/version-catalog

oguzkocer committed:
Use version catalog for wordpress-utils

oguzkocer committed:
Update dependency-catalog version to 1.0.0

ParaskP7 reviewed (commented):
👋 @oguzkocer !

I have reviewed and tested this PR as per the instructions. Everything LGTM, awesome work! 🌟  💯 🥇 

Everything looks and works as per your PR description, especially the WCAndroid vs. FluxC versions and the fact that you chose those newer ones from WCAndroid to be added to the `Version Catalog` but overrode those on the FluxC side afterwards.

I have nothing to add or suggest for this PR. I'll post any questions I have about `Version Catalog` in general on the accompanying P2 post instead. As such it is a 👍 from my side. Just let me know when you mark this PR ready for review so I can take another look before approving.

oguzkocer committed:
Merge remote-tracking branch 'origin/trunk' into add/version-catalog

oguzkocer committed:
Update catalogVersion to 1.2.0 and remove dagger version override

oguzkocer commented:
@ParaskP7 I updated this PR with current `trunk`, resolved the conflicts and marked it as ready for review.

The conflicts were about the version changes for AGP & Dagger. I updated both in https://github.com/Automattic/android-dependency-catalog/pull/3 and updated the catalog version 93b7aa6dedcfa5fdf312aa29bac95fdebaab19ab. I've then compared the dependencies with current `trunk` and verified that this PR is not changing any dependency versions.

Could you take another look at it so we can merge it? I think we have given more than enough time to it since the announcement.

ParaskP7 commented:
👋 @oguzkocer !

> @ParaskP7 I updated this PR with current trunk, resolved the conflicts and marked it as ready for review.

Great, thanks for doing that! 🥇 

> The conflicts were about the version changes for AGP & Dagger. I updated both in https://github.com/Automattic/android-dependency-catalog/pull/3 and updated the catalog version https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/93b7aa6dedcfa5fdf312aa29bac95fdebaab19ab. I've then compared the dependencies with current trunk and verified that this PR is not changing any dependency versions.

Awesome, good job! 💯 

> Could you take another look at it so we can merge it?

I did take another look and everything seems as expected to me. As you noted, this change is all about `AGP` & `Dagger`, it is another 👍 from my side, we can definitely proceed with merging! 🟢 

> I think we have given more than enough time to it since the announcement.

Indeed! 📢
